feat: Создать интерфейсы FilmStorage и UserStorage, в которых будут о…

RodionovRL opened the pull request:
…пределены методы добавления, удаления и модификации объектов.

Создайть классы InMemoryFilmStorage и InMemoryUserStorage, имплементирующие новые интерфейсы, перенести туда всю логику хранения, обновления и поиска объектов. Создать UserService, который будет отвечать за операции с пользователями Создать FilmService, который будет отвечать за операции с фильмами Отредактировать контроллеры для нового функционала

RodionovRL committed:
feat: Создать интерфейсы FilmStorage и UserStorage, в которых будут определены методы добавления, удаления и модификации объектов.
Создайть классы InMemoryFilmStorage и InMemoryUserStorage, имплементирующие новые интерфейсы, перенести туда всю логику хранения, обновления и поиска объектов.
Создать UserService, который будет отвечать за операции с пользователями
Создать FilmService, который будет отвечать за операции с фильмами
Отредактировать контроллеры для нового функционала

RodionovRL committed:
style: code style

RodionovRL committed:
style: code style

RodionovRL committed:
fix: add getter in ErrorResponse.java

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,29 @@
+import ru.yandex.practicum.filmorate.exception.UserNotFoundException;
+
+@RestControllerAdvice
+public class ErrorHandler {
```
В каждый из методов обработчика исключений имеет смысл добавить логгирование происходящего (уровень debug или info). Это полезно для последующего анализа работы приложения. Если это не делать, сообщения о том, что что-то пошло не так, будет получать только пользователь, но не разработчик.
В нынешней реализации эта проблема решается так, что перед каждым throw new UserNotFoundException/FilmNotFoundException логгирование делается вручную. Так тоже можно, но нужно постоянно об этом помнить и дублировать сообщения в тексте исключения и в логах

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/api/UserStorage.java:
```diff
@@ -0,0 +1,21 @@
+
+    User replace(int id, User user);
+
+    Collection<User> values();
```
лучше дать методу какое-нибудь более осмысленное название (getAllUsers(), например), чтобы было понятно что _именно_ за значения он возвращает

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -43,4 +43,54 @@ public ResponseEntity<Collection<User>> getAllUsers() {
+    public ResponseEntity<User> deleteFriend(@PathVariable("id") Integer id,
+                                             @PathVariable("friendId") Integer friendId
+    ) {
+        log.info("получен запрос на удаление друга с id {} у пользователя пользователю с id {} ", friendId, id);
```
как-то странно звучит " у пользователя пользователю с id "...
log.info("запрашивается удаление пользователя {} из списка друзей пользователя {} ", friendId, id);

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,22 +1,30 @@
+        this.userService = userService;
+    }
 
     public Film addFilm(@NotNull Film film) {
```
Здесь @<!-- -->NotNull в значительной степени бесполезен. Ведь метод вызывается из контроллера для объекта, прошедшего валидацию

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -29,10 +37,6 @@ public Film addFilm(@NotNull Film film) {
 
     public Film updateFilm(@NotNull Film film) {
```
аналогично addFilm

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,21 +1,28 @@
+        this.users = users;
+    }
 
     public User addUser(@NotNull User newUser) {
```
про @<!-- -->NotNull аналогично сервису фильмов

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -29,12 +36,12 @@ public User addUser(@NotNull User newUser) {
 
     public User updateUser(@NotNull User user) {
 
-        if (!users.containsKey(user.getId())) {
```
аналогично updateFilm

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public List<User> getUsersByIds(Set<Integer> friendsIds) {
+        if (friendsIds == null) {
```
разве такое возможно? Список друзей получается из User.getFriendsIds(), а это геттер к не-null полю

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -20,11 +24,27 @@ public class User {
     @Past(message = "Пользователь ещё не родился?")
     private LocalDate birthday;
+
+    private Set<Integer> friendsIds = new HashSet<>();
```
поле можно сделать финальным, поскольку может меняться только содержимое коллекции, но не ссылка не неё

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -29,12 +36,12 @@ public User addUser(@NotNull User newUser) {
-        }
         checkName(user);
         User oldUser = users.replace(user.getId(), user);
+        if (oldUser == null) {
```
null, который вернёт replace, означает, что раньше (т.е. _до_ обновления) этому ключу не соответствовало значений. Но обновление тем не менее произойдёт и в мапе хранилища появится запись для нового пользователя. Бросать в этой ситуации исключение неправильно.
По хорошему, эта проверка - что пользователь существует - должна делаться на уровне хранилища (представьте, что в качестве хранилища выступает БД) и если пользователь не найден, то исключение должно бросаться оттуда.

В этом смысле организация взаимодействия сервиса и хранилища фильмов организовано более корректно

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -44,10 +48,49 @@ public Collection<Film> getAllFilms() {
 
+    public Film getFilmById(int id) {
+        Film film = films.getFilmById(id);
+        if (film == null) {
```
эту проверку лучше перенести на уровень хранилища

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,22 +1,30 @@
 @Service
 public class FilmService {
+
     private static int ids;
```
генерированием идентификаторов и их присваиванием должно заниматься хранилище - это его функция, а не бизнес-слоя, реализуемого сервисом. В перспективе в качестве хранилища будет выступать БД и идентификаторы будут генерироваться ею. В нынешней in-memory реализации счетчик фильмов должен размещаться в FilmStorage (пользователей - в UserStorage)

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,22 +1,30 @@
-    @Getter
-    private final Map<Integer, Film> films = new HashMap<>();
+    FilmStorage films;
+    UserService userService;
```
Сервисы не должны знать друг о друге. Это нормально, что сервис фильмов должен получать информацию о пользователях, но для этого нужно использовать методы UserStorage, а не UserService

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,21 +1,28 @@
 @Service
 public class UserService {
 
     private static int ids;
```
аналогично FilmService.ids

kvach2avl reviewed (commented):
Вопросы, замечания и рекомендации приводятся ниже

RodionovRL committed:
fix: перенести логику добавления id Storage
fix: перенести проверки наличия по id в Storage
style: дать методам осознанные имена
fix: добавить логи в обработчики исключений
fix: проверит и исправить текст логов
fix: избавиться от лишних @NotNull
fix: финализировать поля Set<> в моделях
fix: отвязать сервисы друг от друга - использовать хранилище

RodionovRL committed:
fix: исправить вызов конструктора Film и User после final полей Set

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,53 +1,87 @@
-        int newId = ++ids;
+    public Film getFilmById(int id) {
+        Film film = filmStorage.getFilmById(id);
+        if (film == null) {
```
теперь это можно не проверять - film никогда не будет null, иначе filmStorage.getFilmById бросит исключение

RodionovRL replied:
Видимо, невнимательно вычистил проверки. Исправлю.

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,53 +1,87 @@
+    }
+
+    public Film updateFilm(Film film) {
+
```
зачем здесь пустая строка?

RodionovRL replied:
По пустым строкам замечание понял. Исправлюсь.

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,53 +1,87 @@
 
-    public Film updateFilm(@NotNull Film film) {
+    public Film addFilm(Film film) {
 
```
зачем здесь пустая строка?

kvach2avl commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,53 +1,87 @@
     }
 
     public Collection<Film> getAllFilms() {
 
```
зачем здесь пустая строка?

kvach2avl reviewed (commented):
ok
Общая рекомендация: не стоит отделять название метода от его реализации пустой строкой
